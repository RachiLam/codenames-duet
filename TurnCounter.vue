<template>
  <div
    class="turn-counter counter-like"
    :class="type"
    @click="$emit('click')"
  />
</template>

<script>
export default {
  props: {
    type: {
      type: String,
      required: true,
    },
  },
};
</script>

<style lang="scss" scoped>
  @import './variables';

  .turn-counter {
    &.player {
      background-color: $primary;
    }

    &.remaining {
      background-color: $accent-dark;
    }

    &.error {
      background-color: $error;
    }
  }
</style>
