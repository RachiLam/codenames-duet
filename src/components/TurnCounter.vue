<template>
  <div
    class="turn-counter counter-like"
    :class="type"
    @click="$emit('click')"
  />
</template>

<script>
export default {
  props: {
    type: {
      type: String,
      required: true,
      validator: (value) => ['error', 'normal', 'player'].includes(value),
    },
  },
};
</script>

<style lang="scss" scoped>
  @use 'styles/colors';

  .turn-counter {
    &.error {
      background-color: colors.$error;
    }

    &.normal {
      background-color: colors.$accent-dark;
    }

    &.player {
      background-color: colors.$success;
    }
  }
</style>
