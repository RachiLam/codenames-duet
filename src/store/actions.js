export const actions = {
  reload: () => {
    window.location.reload();
  },
};
