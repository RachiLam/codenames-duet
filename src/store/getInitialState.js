import { utils } from '../utils';

const makeCardState = (id) => ({
  id,
  word: utils.getWord(),
  status: [null, null],
});

const makePlayerState = (id) => ({
  id,
  successfulTurnCount: 0,
});

export const getInitialState = () => ({
  cards: utils.range(25).map((index) => makeCardState(index)),
  players: utils.range(2).map((index) => makePlayerState(index)),
  sideboard: {
    maxTurns: 9,
  },
  flags: {
    adminBoard: false,
  },
  adminMode: false,
});
