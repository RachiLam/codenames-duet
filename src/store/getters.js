export const getters = {
  turnsTaken: (state) => {
    const successfulTurns = state.players.reduce(
      (turnsTaken, player) => turnsTaken + player.successfulTurnCount,
      0,
    );

    const failedTurns = state.cards.reduce(
      (turnsTaken, card) => (
        turnsTaken
        + (card.status[0] === 'bystander' ? 1 : 0)
        + (card.status[1] === 'bystander' ? 1 : 0)
      ),
      0,
    );

    return successfulTurns + failedTurns;
  },
};
