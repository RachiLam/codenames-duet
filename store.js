import Vue from 'vue';
import Vuex from 'vuex';
import dictionary from './dictionary';

const range = (count) => Array.from({ length: count }).map((value, index) => index);

const getWord = () => {
  const wordIndex = Math.floor(Math.random() * dictionary.length);
  return dictionary[wordIndex];
};

const makeCardState = (id) => ({
  id,
  word: getWord().toUpperCase(),
  isAssassin: [false, false],
  isAgent: [false, false],
  isBystander: [false, false],
});

const makePlayerState = (id) => ({
  id,
  successfulTurnCount: 0,
});

export default new Vuex.Store({
  state: {
    cards: range(25).map((index) => makeCardState(index)),
    players: range(2).map((index) => makePlayerState(index)),
    sideboard: {
      maxTurns: 9,
    },
  },
  getters: {
    turnsTaken: (state) => {
      const successfulTurns = state.players.reduce(
        (turnsTaken, player) => turnsTaken + player.successfulTurnCount,
        0,
      );

      const failedTurns = state.cards.reduce(
        (turnsTaken, card) => {
          const player0TurnsTaken = card.isBystander[0] ? 1 : 0;
          const player1TurnsTaken = card.isBystander[1] ? 1 : 0;

          return turnsTaken + player0TurnsTaken + player1TurnsTaken;
        },
        0,
      );

      return successfulTurns + failedTurns;
    },
  },
  mutations: {
    /* eslint-disable no-param-reassign */
    updateSuccessfulTurnCount(state, { playerId, count }) {
      state.players[playerId].successfulTurnCount = count;
    },
    toggleCardState(state, { cardId, key, playerId }) {
      Vue.set(state.cards[cardId][key], playerId, !state.cards[cardId][key][playerId]);
    },
    /* eslint-enable no-param-reassign */
  },
});
