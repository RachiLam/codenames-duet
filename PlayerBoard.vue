<template>
  <div class="player-board">
    <button
      class="add-counter counter-like"
      @click="addSuccessCounter"
    />
    <turn-counter
      v-for="index in player.successfulTurnCount"
      :key="`player-turn-counter-${player.id}${index}`"
      type="player"
      @click="removeSuccessCounter"
    />
  </div>
</template>

<script>
import { mapState, mapMutations } from 'vuex';
import TurnCounter from './TurnCounter.vue';

export default {
  components: {
    TurnCounter,
  },
  props: {
    playerId: {
      type: Number,
      required: true,
    },
  },
  computed: {
    ...mapState(['players']),
    player() {
      return this.players[this.playerId];
    },
  },
  methods: {
    ...mapMutations([
      'updateSuccessfulTurnCount',
    ]),
    addSuccessCounter() {
      this.updateSuccessfulTurnCount({
        playerId: this.player.id,
        count: this.player.successfulTurnCount + 1,
      });
    },
    removeSuccessCounter() {
      this.updateSuccessfulTurnCount({
        playerId: this.player.id,
        count: this.player.successfulTurnCount - 1,
      });
    },
  },
};
</script>

<style lang="scss" scoped>
  @import 'variables.scss';

  .player-board {
    display: flex;
    align-items: center;

    $starting-edge-offset: 400px;
    &.player-board-0 {
      flex-direction: row-reverse;
      padding-right: $starting-edge-offset;
    }

    &.player-board-1 {
      padding-left: $starting-edge-offset;
    }

    .add-counter {
      background-color: $accent-light;
      border-radius: 100%;
    }
  }
</style>
