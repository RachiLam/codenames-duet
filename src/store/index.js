import Vuex from 'vuex';
import { getInitialState } from './getInitialState';
import { getters } from './getters';
import { mutations } from './mutations';
import { actions } from './actions';

export default new Vuex.Store({
  state: getInitialState(),
  getters,
  mutations,
  actions,
});
