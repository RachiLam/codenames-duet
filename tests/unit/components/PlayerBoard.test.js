import { getInitialState } from 'store/getInitialState';
import PlayerBoard from 'components/PlayerBoard.vue';
import TurnCounter from 'components/TurnCounter.vue';

const { mount, shallowMount } = getMountUtils({
  component: PlayerBoard,
  defaultProps: { playerId: 0 },
});

const getAddCounterButton = (wrapper) => wrapper.find('button.add-counter');

describe('components/PlayerBoard', () => {
  context('always', () => {
    it('renders an "add counter" button', () => {
      const wrapper = shallowMount();
      expect(getAddCounterButton(wrapper).element).toMatchSnapshot();
    });
  });

  context('at the beginning of the game', () => {
    it('does not render turn counters', () => {
      const wrapper = shallowMount({
        flatStore: getInitialState(),
      });

      const turnCounters = wrapper.findAllComponents(TurnCounter).wrappers;
      expect(turnCounters.length).toBe(0);
    });
  });

  context('when the "add counter" button is pressed', () => {
    let wrapper;
    let addCounterButton;
    const updateSuccessfulTurnCount = jest.fn();
    beforeAll(() => {
      wrapper = shallowMount({
        props: { playerId: 1 },
        flatStore: {
          players: [
            undefined,
            { successfulTurnCount: 5 },
          ],
          updateSuccessfulTurnCount,
        },
      });

      addCounterButton = getAddCounterButton(wrapper);
      addCounterButton.trigger('click');
    });

    it('increments the players turn count', () => {
      expect(updateSuccessfulTurnCount).toHaveBeenCalledWith(
        expect.anything(),
        { playerId: 1, count: 6 },
      );
    });
  });

  context('when the player has taken turns', () => {
    let wrapper;
    let turnCounters;
    beforeAll(() => {
      wrapper = shallowMount({
        props: { playerId: 0 },
        flatStore: {
          players: [
            { successfulTurnCount: 2 },
          ],
        },
      });

      turnCounters = wrapper.findAllComponents(TurnCounter).wrappers;
    });

    it('renders a turn counter for each successful turn', () => {
      expect(turnCounters.length).toBe(2);
    });

    it('renders "player" turn counters', () => {
      const elements = turnCounters.map(({ element }) => element);
      expect(elements).toMatchSnapshot();
    });
  });

  context('when a turn counter is pressed', () => {
    let wrapper;
    let secondTurnCounter;
    const updateSuccessfulTurnCount = jest.fn();
    beforeAll(() => {
      wrapper = mount({
        props: { playerId: 0 },
        flatStore: {
          players: [
            { successfulTurnCount: 3 },
          ],
          updateSuccessfulTurnCount,
        },
      });

      [, secondTurnCounter] = wrapper.findAllComponents(TurnCounter).wrappers;
      secondTurnCounter.trigger('click');
    });

    it('decrements the players turn count', () => {
      expect(updateSuccessfulTurnCount).toHaveBeenCalledWith(
        expect.anything(),
        { playerId: 0, count: 2 },
      );
    });
  });
});
