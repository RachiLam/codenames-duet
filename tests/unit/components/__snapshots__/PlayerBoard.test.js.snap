// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/PlayerBoard always renders an "add counter" button 1`] = `
<button
  class="add-counter counter-like"
/>
`;

exports[`components/PlayerBoard when the player has taken turns renders "player" turn counters 1`] = `
Array [
  <turn-counter-stub
    type="player"
  />,
  <turn-counter-stub
    type="player"
  />,
]
`;
